server/src/main/java/org/elasticsearch/indices/IndicesService.java

```diff
@@ -457,7 +457,24 @@ public NodeIndicesStats stats(CommonStatsFlags flags) {
             }
         }
 
-        return new NodeIndicesStats(commonStats, statsByShard(this, flags));
+        return new NodeIndicesStats(commonStats, statsByIndex(this, flags), statsByShard(this, flags));
+    }
+
+    static Map<Index, CommonStats> statsByIndex(final IndicesService indicesService, final CommonStatsFlags flags) {
+        final Map<Index, CommonStats> statsByIndex = new HashMap<>();
```

Comment:
Nit: Might be worth pre-sizing this via `Maps.newHashmapWithExpectedSize` for the large node case

Reply:
Good idea!